File: examples/oshmem_circular_shift.c
/*
 * Copyright (c) 2014-2016 Mellanox Technologies, Inc.
 *                         All rights reserved.
 * $COPYRIGHT$
 *
 * Additional copyrights may follow
 *
 * $HEADER$
 */

#include <stdio.h>
#include <shmem.h>

int main (void)
{
    static int aaa, bbb;
    int num_pes, my_pe, peer;

    shmem_init();

    num_pes = shmem_n_pes();
    my_pe = shmem_my_pe();

    peer = (my_pe + 1) % num_pes;

    printf("Process %d gets message from %d (%d processes in ring)\n", my_pe, peer, num_pes);
    shmem_int_get(&aaa, &bbb, 1, peer);

    shmem_barrier_all();
    printf("Process %d exiting\n", my_pe);
    shmem_finalize();

    return 0;
}

File: examples/oshmem_circular_shift.java
import shmem.*;

public class oshmem_circular_shift
{
    public static void main(String[] args) throws ShMemException
    {
        ShMem.startPEs(0);

        int numPEs = ShMem.getNumPEs(),
            myPE   = ShMem.getMyPE(),
            peer   = (myPE + 1) % numPEs;

        int[] aaa = new int[1];
        Addr  bbb = new Addr(4);

        System.out.println("Process "+ myPE +" gets message from "+
                           peer +" ("+ numPEs +" processes in ring)");

        bbb.getInt(aaa, peer);
        ShMem.barrierAll();
        bbb.free();
        System.out.println("Process "+ myPE +" exiting");
    }
}
